qiskit_optimization/applications/ising/tsp.py

```diff
@@ -1,274 +1,110 @@
-            then visit no where as the third city).
-    """
-    n = int(np.sqrt(len(x)))
-    z = []
-    for p__ in range(n):
-        p_th_step = []
+    def _build_quadratic_program(self):
+        mdl = Model(name='tsp')
+        n = self._g.number_of_nodes()
+        x = {(i, p): mdl.binary_var(name='x_{0}_{1}'.format(i, p))
+             for i in range(n) for p in range(n)}
+        tsp_func = mdl.sum(self._g.edges[i, j]['weight'] * x[(i, p)] * x[(j, (p+1) % n)]
+                           for i in range(n) for j in range(n) for p in range(n) if i != j)
+        mdl.minimize(tsp_func)
         for i in range(n):
-            if x[i * n + p__] >= 0.999:
-                p_th_step.append(i)
-        if len(p_th_step) == 1:
-            z.extend(p_th_step)
-        else:
-            z.append(p_th_step)
-    return z
+            mdl.add_constraint(mdl.sum(x[(i, p)] for p in range(n)) == 1)
+        for p in range(n):
+            mdl.add_constraint(mdl.sum(x[(i, p)] for i in range(n)) == 1)
+        qp = QuadraticProgram()
+        qp.from_docplex(mdl)
+        return qp
+
+    def plot_graph(self, x, pos=None):
+        route = self.interpret(x)
+        nx.draw(self._g, with_labels=True, pos=pos)
+        nx.draw_networkx_edges(
+            self._g,
+            pos,
+            edgelist=[(route[i], route[(i+1) % len(route)]) for i in range(len(route))],
+            width=8, alpha=0.5, edge_color="tab:red",
+            )
+
+    def interpret(self, x):
+        n = int(np.sqrt(len(x)))
+        route = []
+        for p__ in range(n):
+            for i in range(n):
+                if x[i * n + p__]:
+                    route.append(i)
+        return route
+
+    @staticmethod
+    def random_graph(n=5, low=0, high=100, seed=None):
+        random.seed(seed)
+        pos = {i: (random.randint(low, high), random.randint(low, high)) for i in range(n)}
+        g = nx.random_geometric_graph(n, np.hypot(high-low, high-low)+1, pos=pos)
+        for u, v in g.edges:
+            delta = [g.nodes[u]['pos'][i] - g.nodes[v]['pos'][i] for i in range(2)]
+            g.edges[u, v]['weight'] = np.rint(np.hypot(delta[0], delta[1]))
+        return g
+
+    @staticmethod
+    def parse_tsplib_format(filename):
+        """Read graph in TSPLIB format from file.
+
+        Args:
+            filename (str): name of the file.
+
+        Returns:
+            TspData: instance data.
+
+        """
+        name = ''
+        coord = []
+        with open(filename) as infile:
+            coord_section = False
+            for line in infile:
+                if line.startswith('NAME'):
+                    name = line.split(':')[1]
+                    name.strip()
+                elif line.startswith('TYPE'):
+                    typ = line.split(':')[1]
+                    typ.strip()
+                    if typ != 'TSP':
+                        logger.warning('This supports only "TSP" type. Actual: %s', typ)
```

Comment:
I am thinking that perhaps this logger.warning (and the one below) ought to be raising a QiskitOptimizationError rather than logging a warning and continuing. The warning may be missed and it seems in these cases that the file contains something that is not supported by the TSP application here so would be better as an error?